subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,491 @@
+{
+	return 0;
+}
+
+static struct usbd_class_api bt_hci_api = {
+	.request = bt_hci_request,
+	.update = bt_hci_update,
+	.enable = bt_hci_enable,
+	.disable = bt_hci_disable,
+	.control_to_dev = bt_hci_ctd,
+	.init = bt_hci_init,
+};
+
+#define BT_HCI_DESCRIPTOR_DEFINE(n)						\
+static struct usbd_bt_hci_desc bt_hci_desc_##n = {				\
+	.if0 = {								\
+		.bLength = sizeof(struct usb_if_descriptor),			\
+		.bDescriptorType = USB_DESC_INTERFACE,				\
+		.bInterfaceNumber = 0,						\
+		.bAlternateSetting = 0,						\
+		.bNumEndpoints = 3,						\
+		.bInterfaceClass = USB_BCC_WIRELESS_CONTROLLER,			\
+		.bInterfaceSubClass = USB_BT_HCI_SUBCLASS,			\
+		.bInterfaceProtocol = USB_BT_HCI_PROTOCOL,			\
+		.iInterface = 0,						\
+	},									\
+										\
+	.if0_int_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x81,					\
+		.bmAttributes = USB_EP_TYPE_INTERRUPT,				\
+		.wMaxPacketSize = sys_cpu_to_le16(BT_HCI_DEFAULT_INT_EP_MPS),	\
+		.bInterval = 1,							\
+	},									\
+										\
+	.if0_out_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x02,					\
+		.bmAttributes = USB_EP_TYPE_BULK,				\
+		.wMaxPacketSize = sys_cpu_to_le16(BT_HCI_DEFAULT_BULK_EP_MPS),	\
+		.bInterval = 0,							\
+	},									\
+										\
+	.if0_in_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x82,					\
+		.bmAttributes = USB_EP_TYPE_BULK,				\
+		.wMaxPacketSize = sys_cpu_to_le16(BT_HCI_DEFAULT_BULK_EP_MPS),	\
+		.bInterval = 0,							\
+	},									\
+};
+
+#define BT_HCI_CLASS_DATA_DEFINE(n)						\
+	static struct bt_hci_data bt_hci_data_##n = {				\
+		.sync_sem = Z_SEM_INITIALIZER(bt_hci_data_##n.sync_sem, 0, 1),	\
+	};									\
+										\
+	static struct usbd_class_data bt_hci_class_##n = {			\
+		.desc = (struct usb_desc_header *)&bt_hci_desc_##n,		\
+		.priv = &bt_hci_data_##n,					\
+		.v_reqs = &bt_hci_vregs,					\
+	};									\
+										\
+	USBD_DEFINE_CLASS(bt_hci_##n, &bt_hci_api, &bt_hci_class_##n);
+
+/*
+ * Bluetooth subsystem does not support multiple HCI instances,
+ * but we are almost ready for it.
+ */
+BT_HCI_DESCRIPTOR_DEFINE(0)
+BT_HCI_CLASS_DATA_DEFINE(0)
+
+static int bt_hci_preinit(const struct device *dev)
+{
+	int ret;
+
+	ret = bt_enable_raw(&bt_hci_tx_queue);
+	if (ret) {
+		LOG_ERR("Failed to open Bluetooth raw channel: %d", ret);
+		return ret;
+	}
+
+	k_thread_create(&rx_thread_data, rx_thread_stack,
+			K_KERNEL_STACK_SIZEOF(rx_thread_stack),
+			bt_hci_rx_thread, NULL, NULL, NULL,
+			K_PRIO_COOP(8), 0, K_NO_WAIT);
+
+	k_thread_name_set(&rx_thread_data, "bt_hci_rx");
+
+	k_thread_create(&tx_thread_data, tx_thread_stack,
+			K_KERNEL_STACK_SIZEOF(tx_thread_stack),
+			bt_hci_tx_thread, (void *)&bt_hci_0, NULL, NULL,
+			K_PRIO_COOP(8), 0, K_NO_WAIT);
```

Comment:
Is the priority supposed to be 8 and never anything else? Should it be put into a Kconfig option?

Reply:
Yes, I am not sure we need both threads, I have to check wath context bt_send() expects.